enhancements/okd-on-fedora-coreos.md

```diff
@@ -0,0 +1,159 @@
+
+One benefit of this is that version drift between MCD from the boot image and from the current machine-os-content is ruled out, as the MCD will always come from the MCO image from the current payload.
+
+Note: MCO images are currently UBI7 based. For RHEL CoreOS it should be preferred to pull an MCD built inside a UBI8 container.
+
+#### Dual Ignition config spec v2/v3 support in Machine Config Operator 
+
+The proposal is to add dual Ignition support to the MCO so that OKD does not require a forked version and can instead be built from the canonical branches in the machine-config-operator repository.
+
+In order to achive this, the following will be done:
+
+Firstly, the ability to parse and translate v3.x configs to v2.2 will be added. This way the MCO will be able to process v3.x config that is either provided by the installer during install, or by the cluster admin at a later point. The rendered machine config representing the current on-disk state internally will remain spec v2.2 for now.
+
+Secondly, the Machine Config Server will gain the ability to inspect the User Agent HTTP header of the incoming config request to serve a spec version supported by the requesting system. For this, the canonical rendered machine config will be translated up to spec v3 if the User Agent shows an Ignition version larger than 2.0.0. 
+
+This is implemented by this PR: https://github.com/openshift/machine-config-operator/pull/1703
+
+The last commit in that PR already shows that the spec used to store on-disk state internally (i.e. the rendered machine config) could already be switched over to v3 without any breaking changes in external APIs. This however is not required to enable using the same MCO images for OKD and OCP.
+
+### 3: Sample Operators Changes
+
+#### Mark RHEL based operator images as optional
+
+The samples operator degrades on OKD due to the fact that not all images are pullable without a valid pull secret. The behaviour should change so it does not degrade on OKD clusters for users that have not specified a valid pull secret.
+
+Alternatively, all RHEL based images this concers could be rebuilt on a CentOS or Fedora base, and a mechanism to pull them in case of no valid pull secret could be implemented in `cluster-samples-operator`. 
+
+### 4: Installer Changes
+
+These changes are optional for the immediate goal of releasing OKD 4 GA. They should however be strongly considered in order to facilitate dropping the OKD fork of installer rather sooner than later.
+
+#### Add fcos.json
+
+Support for an `fcos.json` to exist alongside `rhcos.json` for OS image references.
+
+This could be enabled either via a build-time flag to produce seperate OKD installer artifacts, or an `fcos` or `okd` run-time flag to pass to the installer for it to provision Fedora CoreOS instead of RHEL CoreOS hosts (in that case it would default to contents of `rhcos.json`). 
```

Comment:
I'd prefer the build-time flag, similar to what libvirt approach:

* [ ] Rework existing installer to have all ignition operators in a single file (https://github.com/openshift/installer/pull/3248)
* [ ] Create Ignition 3 implementation of the same functions, hidden under `okd` build flag
* [ ] Build `installer:okd` image with this flag enabled
* [ ] Add `isOKD` internal setting and conditionally enable okd-specific features:
  * use `fcos.json` instead of `rhcos.json`
  * allow empty pull secret
  * restrict to community operators
* [ ] Enable libvirt provider in OKD installer

Reply:
+1 from me. @abhinavdahiya, WDYT?

Reply:
Why not rename rhcos.json to machine-os.json and make that the build time replacement instead?

Branches on “is okd” raise red flags for me - the imply OKD isn’t the same source.

The restriction to community operators can be handled dynamically in many cases (does the payload include the operator).

The pull secret is going to be relevant for some deployments no matter what - perhaps instead of making the check conditional we instead check whether the pull of the release image requires auth as a gate during pre validation (which has the side benefit of verifying the credentials)

Reply:
Bikeshed: `bootimage.json`

Reply:
> Why not rename rhcos.json to machine-os.json and make that the build time replacement instead?

Sounds good to me

> Branches on “is okd” raise red flags for me - the imply OKD isn’t the same source.

We still need to branch on few options here - i.e. default SDN, Ignition v3 etc. `IsOKD` is not exposed externally, cannot be changed via config and being set on build time

> The restriction to community operators can be handled dynamically in many cases (does the payload include the operator).

So instead of installer injecting operators we'd have to have a setting other operators would read? 

> we instead check whether the pull of the release image requires auth as a gate during pre validation

This check would happen in the installer, right?